# 1장 파이썬 머신러닝 이해  



[머신러닝 완벽 가이드] 책을 기반으로 제가 정리하고 요약한 내용입니다.  
1장은 머신러닝의 기초인 넘파이와 판다스에 대한 개념입니다.

## 넘파이
넘파이는 배열 기반의 연산은 물론이고 다양한 데이터 핸들링 기능을 제공합니다.

In [1]:
import numpy as np # 넘파이 모듈 임포트

### ndarray 데이터 타입

In [2]:
array1 = np.array([1,2,3]) 
# 넘파이의 array()함수는 파이썬의 리스트와 같은 인자를 ndarray로 변환하는 기능 수행
print('array1 type:',type(array1))
print('array1 type 형태:',array1.shape)

array2 = np.array([[1,2,3],
                   [4,5,6]])
print('array2 type:',type(array2))
print('array2 type 형태:',array2.shape)

array1 type: <class 'numpy.ndarray'>
array1 type 형태: (3,)
array2 type: <class 'numpy.ndarray'>
array2 type 형태: (2, 3)


In [3]:
list1 = [1,2,3]
print(type(list1))
array1 = np.array(list1)
print(type(array1))
print(array1.dtype) #ndarray내의 데이터타입은 .dtype 으로 확인 가능

<class 'list'>
<class 'numpy.ndarray'>
int32


In [4]:
list2 = [1,2,'test']
array2 = np.array(list2)
print(array2.dtype)

list3 = [1,2,3.1]
array3 = np.array(list3)
print(array3.dtype) 

<U11
float64


다른 데이터 유형이 ndarray로 변경되면 데이터 크기가 더 큰 데이터 타입으로 형 변환이 적용됩니다.

In [5]:
list3 = [1, 2, 3]
array4 =np.array(list3)
print(array4.dtype)
array4 =array4.astype(float)
print(array4.dtype)

int32
float64


ndarray 내 데이터값의 타입 변경은 astype()을 이용해 가능합니다.

### ndarray 생성

In [6]:
arange = np.arange(10)
print(arange)
print(arange.dtype)
print(arange.shape)

[0 1 2 3 4 5 6 7 8 9]
int32
(10,)


arange()는 0부터 함수 인자 값 -1까지의 값을 순차적으로 ndarray의 데이터값으로 변환해줍니다.  
start, stop을 지정해서 원하는 값부터 시작한 연속 값을 부여할 수 있습니다.

In [7]:
zeros = np.zeros((3,2))
print(zeros)
print(zeros.dtype)
print(zeros.shape)

[[0. 0.]
 [0. 0.]
 [0. 0.]]
float64
(3, 2)


zeros()는 함수 인자로 튜플 형태의 shape값을 입력하면 모든 값을 0으로 채운 해당 shape를 가진 ndarray를 반환합니다. 이와 유사하게 ones()는 모든 값을 1로 채운 ndarray를 반환합니다. default는 float입니다.

In [8]:
arange = np.arange(10)
print(arange)
print(arange.reshape(2,5))
print(arange.reshape(5,-1))

[0 1 2 3 4 5 6 7 8 9]
[[0 1 2 3 4]
 [5 6 7 8 9]]
[[0 1]
 [2 3]
 [4 5]
 [6 7]
 [8 9]]


reshape()는 원하는 형태의 ndarray로 변환해주며, 지정된 사이즈로 변경이 불가능하면 오류가 발생합니다.  
-1를 인자로 사용하면 원래 ndarray와 호환되는 새로운 shape로 변환해줍니다.

### ndarray 데이터 세트 선택 - 인덱싱

In [9]:
array1 = np.arange(start=1,stop=10)
array1

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [10]:
array1[2]
print(array1[2],array1[2].dtype)
array1[:5]
print(array1[array1[:5]])
array1[:-1]
print(array1[:-1])

3 int32
[2 3 4 5 6]
[1 2 3 4 5 6 7 8]


인덱스는 0부터 시작하므로 [2]는 3번째 인덱스 위치의 데이터 값을 의미합니다.  
타입은 ndarray타입이 아니고 ndarray 내의 데이터값을 의미합니다.  
-1은 맨 뒤 데이터값을 의미하며 -2는 맨 뒤에서 두 번째에 있는 데이터값을 말합니다.

In [11]:
array1d = np.arange(start=1,stop=10)
array2d = array1d.reshape(3,3)
print(array2d)
print(array2d[0,0])
print(array2d[1,1])
print(array2d[2,2])

[[1 2 3]
 [4 5 6]
 [7 8 9]]
1
5
9


다차원 ndarray에서 단일 값 추출 경우에는 콤마(,)로 분리된 로우와 칼럼 위치의 인덱스를 통해 접근합니다.

In [12]:
array = np.arange(start=1,stop=10)
print(array)
print(array[:3])
print(array[3:])
print(array[:-1])

[1 2 3 4 5 6 7 8 9]
[1 2 3]
[4 5 6 7 8 9]
[1 2 3 4 5 6 7 8]


':' 기호 앞에 시작 인덱스를 생략하면 자동으로 맨 처음 인덱스인 0으로 간주
':' 기호 뒤에 종료 인덱스를 생략하면 자동으로 맨 마지막 인덱스로 간주
':' 기호 앞/뒤에 시작/종료 인덱스를 생략하면 자동으로 맨처음/맨 마지막 인덱스로 간주

In [13]:
array = np.arange(start=1,stop=10)
array = array.reshape(3,3)
print(array)
print(array[0:2,0:2])
# print(array[1:2,1:3])
print(array[1:3,:2])

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1 2]
 [4 5]]
[[4 5]
 [7 8]]


2차원 ndarray에서도 1차원 ndarray에서의 슬라이싱과 유사하며, 콤마(,)로 로우와 칼럼 인덱스를 지칭

In [14]:
array = np.arange(1,10)
array = array.reshape(3,3)
print(array)
print(array[[0,1],2])
print(array[[0,1],:2])


[[1 2 3]
 [4 5 6]
 [7 8 9]]
[3 6]
[[1 2]
 [4 5]]


리스트나 ndarray로 인덱스 집합을 지정하면 해당 위치의 인덱스에 해당하는 ndarray를 반환합니다.

In [15]:
array = np.arange(1,10)
print(array)
print(array>5)
print(array[array>5])

[1 2 3 4 5 6 7 8 9]
[False False False False False  True  True  True  True]
[6 7 8 9]


조건 필터링과 검색을 동시에 할 수 있는 방식을 불린 인덱싱 방식이라고 합니다.

### 행렬의 정렬

In [16]:
array = np.array([3,1,9,5])
print(array)
print(np.sort(array))
print(array)
array1 = array.sort()
print(array1)
print(array)

array = np.array([3,1,9,5])
print("내림차순:",np.sort(array)[::-1])

[3 1 9 5]
[1 3 5 9]
[3 1 9 5]
None
[1 3 5 9]
내림차순: [9 5 3 1]


np.sort()는 원본 행렬에는 영향을 주지않고 정렬된 행렬만을 반환하며,  
ndarray.sort()는 원 행렬 자체를 정렬한 형태로 변환하며, 반환값은 None입니다.
모두 오름차순으로 원소를 정렬하며 내림차순으로 정렬하기 위해서는 [::-1]을 적용합니다.

In [17]:
array = np.array([[3,5],
                  [5,4]])
print(array)
print("axis=0:\n",np.sort(array,axis=0))
print("axis=1:\n",np.sort(array,axis=1))

[[3 5]
 [5 4]]
axis=0:
 [[3 4]
 [5 5]]
axis=1:
 [[3 5]
 [4 5]]


2차원 이상 행렬인 경우 axis 축 값 설정을 통해 로우 방향, 또는 칼럼 방향으로 정렬을 수행할 수 있습니다.

In [18]:
array = np.array([3,1,9,5])
print(array)
print("argsort:\n",np.argsort(array))

[3 1 9 5]
argsort:
 [1 0 3 2]


원본 행렬의 원소에 대한 인덱스를 필요로 할 때 np.argsort를 이용합니다.

### 선형대수 연산 - 행렬 내적과 전치 행렬 구하기

In [19]:
A= np.array([[1,2,3],[4,5,6]])
B= np.array([[7,8],[9,10],[11,12]])

np.dot(A,B)


array([[ 58,  64],
       [139, 154]])

행렬 내적은 행렬 곱이며, 두 행렬의 내적은 np.dot()을 이용해 계산합니다.  
주의해야할 특성으로는 왼쪽 행렬의 열 개수와 오른쪽 행렬의 행 개수가 동일해야 합니다.

In [20]:
print(A)
print(A.T)

[[1 2 3]
 [4 5 6]]
[[1 4]
 [2 5]
 [3 6]]


행렬에서 행과 열의 위치를 교환한 원소로 구성한 행렬을 전치행렬이라하며, transpose 또는 T를 이용해 구할 수 있습니다.

## 판다스

판다스의 핵심 객체는 데이터프레임으로, 여러 개의 행과 열로 이뤄진 2차원 데이터를 담는 데이터 구조체입니다. 데이터프레임을 이해하기 전에 다른 중요 객체인 인덱스와 시리즈를 이해하는 것이 중요하며 인덱스는 개별 데이터를 식별하는 '키'값입니다. 시리는 칼럼이 하나뿐인 데이터 구조체이며, 데이터프레임은 칼럼이 여러 개인 데이터 구조체입니다.

In [21]:
import pandas as pd # 판다스 모듈 임포트

In [22]:
df = pd.read_csv('train.csv')
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

데이터프레임으로 로딩할 수 있는 API로  
read_csv( ), read_table( ), read_fwf( )가 존재하며, csv()와 table()의 차이는 필드 구분 문자가 콤마(,)냐, 탭("\t")이냐의 차이입니다.  
가령 탭으로 필드가 구분돼 있다면 read_csv('파일명', sep='\t')처럼 사용 가능하며, sep를 생략하면 자동으로 콤마로 할당합니다.

read_csv('파일명')에서 '파일명'은 파이썬 실행 파일이 있는 디렉터리와 동일한 디렉터리에 존재해야하며, 같은 디렉터리에 존재하지않는 경우에는 해당 파일에 있는 경로를 입력해주면 불러올 수 있습니다.


In [23]:
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

데이터프레임의 모든 데이터를 한번에 표출하면 메모리가 많이 사용되므로 head()를 이용해 n개의 로우를 반환합니다.default는 5개이며 head(3)은 맨 앞 3개의 로우를 반환합니다.

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


info()를 통해 총 데이터 건수 891개, 컬럼수는 12개인 것을 확인 할 수 있습니다.  
컬럼마다 데이터 형태를 확인할 수 있으며, 어떤 컬럼의 몇개의 널값이 존재하는지 확인할 수 있습니다.  
메모리 사용량도 확인 가능합니다.

In [25]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

- count: 널값이 아닌 데이터 건수  
- mean : 전체 데이터 평균값  
- std : 표준편차  
- min,max : 최소,최대값  
- 25%, 50%, 75% : percentile 값

In [26]:
df['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

칼럼의 값이 어떠한 분포로 구성되어 있는지 확인하기 위해서는 value_counts()를 사용합니다. 지정된 칼럼의 데이터값 건수를 반환하며, 분포도를 확인하기 유용합니다.  
즉 데이터프레임의 [] 연산자 내부에 칼럼명을 입력하면 해당 칼럼에 해당하는 시리즈 객체를 반환합니다.

In [27]:
df.sort_values(by='Pclass', ascending=True)

df[['Name','Age']].sort_values(by='Age')
df[['Name','Age','Pclass']].sort_values(by=['Pclass','Age'])

Name    Age  Pclass
305            Allison, Master. Hudson Trevor   0.92       1
297              Allison, Miss. Helen Loraine   2.00       1
445                 Dodge, Master. Washington   4.00       1
802       Carter, Master. William Thornton II  11.00       1
435                 Carter, Miss. Lucile Polk  14.00       1
..                                        ...    ...     ...
859                          Razi, Mr. Raihed    NaN       3
863         Sage, Miss. Dorothy Edith "Dolly"    NaN       3
868               van Melkebeke, Mr. Philemon    NaN       3
878                        Laleff, Mr. Kristo    NaN       3
888  Johnston, Miss. Catherine Helen "Carrie"    NaN       3

[891 rows x 3 columns]

sort_values() by=정렬컬럼, ascending=True 또는 False로 오름차순/내림차순으로 정렬합니다.

### DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환

**리스트, ndarray에서 DataFrame변환**

In [28]:
import numpy as np

col_name=['col']
list1=[1, 2, 3]
array = np.array(list1)

print(array.shape)
df_list = pd.DataFrame(list1, columns=col_name)
print('1차원 리스트로 만든 DataFrame:\n',df_list)
df_ndarray = pd.DataFrame(array, columns=col_name)
print('1차원 ndarray로 만든 DataFrame:\n',df_ndarray)


(3,)
1차원 리스트로 만든 DataFrame:
    col
0    1
1    2
2    3
1차원 ndarray로 만든 DataFrame:
    col
0    1
1    2
2    3


1차원 형태의 데이터를 기반으로 데이터프레임을 생성하므로 칼럼명이 한 개만 필요합니다.

In [29]:
col_name=['col1','col2','col3']

list = [[1, 2, 3],[4, 5, 6]]
array = np.array(list)

print(array)
df_list = pd.DataFrame(list, columns=col_name)
print(df_list)
df_array = pd.DataFrame(array, columns = col_name)
print(df_array)

[[1 2 3]
 [4 5 6]]
   col1  col2  col3
0     1     2     3
1     4     5     6
   col1  col2  col3
0     1     2     3
1     4     5     6


2차원 형태의 데이터를 기반으로 데이터프레임을 생성하므로 칼럼명이 추가된다.

**딕셔너리(dict)에서 DataFrame변환**

In [30]:
dict = {'col1' : [1, 11], 'col2' : [2, 22], 'co3' : [3, 33]}
dict_df = pd.DataFrame(dict)
dict_df

col1  col2  co3
0     1     2    3
1    11    22   33

딕셔너리를 데이터프레임으로 변환 시에는 딕셔너리의 키는 칼럼명, 딕셔너리의 값은 키에 해당하는 칼럼 데이터로 변환됩니다.  

**DataFrame을 ndarray로 변환**

In [31]:
array3 = dict_df.values
print(array3)
print(type(array3))
print(array3.shape)

[[ 1  2  3]
 [11 22 33]]
<class 'numpy.ndarray'>
(2, 3)


데이터프레임을 넘파이 ndarray로 변환하는 것은 데이터프레임 객체의 values를 이용해 쉽게 할 수 있습니다.

**DataFrame을 리스트와 딕셔너리로 변환**

In [32]:
list3 = dict_df.values.tolist()
print(list3)
print(type(list3))

dict3 = dict_df.to_dict('list')
print(dict3)
print(type(dict3))

[[1, 2, 3], [11, 22, 33]]
<class 'list'>
{'col1': [1, 11], 'col2': [2, 22], 'co3': [3, 33]}
<class 'dict'>


데이터프레임을 리스트로의 변환은 values로 얻은 ndarray에 tolist()를 호출하면 됩니다.  
데이터프레임을 딕셔너리로의 변환은 to_dict()를 호출하며, 'list'를 입력하면 딕셔너리 값이 리스트형으로 변환됩니다.

### 데이터프레임의 컬럼 데이터 세트 생성과 수정

In [33]:
df['Age_0']=0
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0

In [34]:
df['Age_by_10'] = df['Age']*10
df['Family_No'] = df['SibSp'] + df['Parch']+1
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      220.0   
1      0          PC 17599  71.2833   C85        C      0      380.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0   

   Family_No  
0          2  
1          2  
2          1

In [35]:
df['Age_by_10'] = df['Age_by_10'] + 100
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

데이터프레임의 칼럼 데이터 세트 생성과 수정 역시 [ ] 연산자를 이용해 쉽게 할 수 있습니다.

### 데이터프레임 데이터 삭제

In [36]:
drop_df = df.drop('Age_0', axis=1 )
drop_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_No  
0      0         A/5 21171   7.2500   NaN        S      320.0          2  
1      0          PC 17599  71.2833   C85        C      480.0          2  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0          1

데이터프레임에서 데이터의 삭제는 drop() 메서드를 이용합니다.중요 파라미터는 labels, axis, inplace로 axis= 0 은 로우 방향, axis=1은 칼럼 방향 축입니다.

여러개의 컬럼들의 삭제는 drop의 인자로 삭제 컬럼들을 리스트로 입력합니다.

In [37]:
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

In [38]:
drop_result = df.drop(['Age_0', 'Age_by_10', 'Family_No'], axis=1, inplace=True)
print(' inplace=True 로 drop 후 반환된 값:',drop_result)
df.head(3)

 inplace=True 로 drop 후 반환된 값: None


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

drop( )메소드의 inplace인자의 기본값은 False 입니다. 이 경우 drop( )호출을 한 DataFrame은 아무런 영향이 없으며 drop( )호출의 결과가 해당 컬럼이 drop 된 DataFrame을 반환합니다.  
  
  
여러개의 컬럼들의 삭제는 drop의 인자로 삭제 컬럼들을 리스트로 입력합니다. inplace=True 일 경우 호출을 한 DataFrame에 drop이 반영됩니다. 이 때 반환값은 None입니다.

### Index 객체

In [39]:
df = pd.read_csv('train.csv')

indexs = df.index
print(indexs)
print(indexs.values) # index 객체를 실제 값 array로 변환

RangeIndex(start=0, stop=891, step=1)
[  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233
 234 235 236 

데이터프레임, 시리즈에서 인덱스 객체만 추출하려면 DataFrame.index 또는 Series.index 속성을 통해 가능합니다.

In [40]:
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

NameError: name 'indexes' is not defined

Index는 1차원 데이터 입니다.

In [41]:
series_fair = df['Fare']
series_fair.head(5)

0     7.2500
1    71.2833
2     7.9250
3    53.1000
4     8.0500
Name: Fare, dtype: float64

In [42]:
print('Fair Series max 값:', series_fair.max())
print('Fair Series sum 값:', series_fair.sum())
print('sum() Fair Series:', sum(series_fair))
print('Fair Series + 3:\n',(series_fair + 3).head(3) )

Fair Series max 값: 512.3292
Fair Series sum 값: 28693.9493
sum() Fair Series: 28693.949299999967
Fair Series + 3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


Series 객체는 Index 객체를 포함하지만 Series 객체에 연산 함수를 적용할 때 Index는 연산에서 제외됩니다. Index는 오직 식별용으로만 사용됩니다.

In [43]:
reset_df = df.reset_index(inplace=False)
reset_df.head(3)

index  PassengerId  Survived  Pclass  \
0      0            1         0       3   
1      1            2         1       1   
2      2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [44]:
print('### before reset_index ###')
value_counts = df['Pclass'].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입:',type(value_counts))

new_value_counts = value_counts.reset_index(inplace=False)
print('### After reset_index ###')
print(new_value_counts)
print('new_value_counts 객체 변수 타입:',type(new_value_counts))

### before reset_index ###
3    491
1    216
2    184
Name: Pclass, dtype: int64
value_counts 객체 변수 타입: <class 'pandas.core.series.Series'>
### After reset_index ###
   index  Pclass
0      3     491
1      1     216
2      2     184
new_value_counts 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


DataFrame 및 Series에 reset_index( ) 메서드를 수행하면 새롭게 인덱스를 연속 숫자 형으로 할당하며 기존 인덱스는 ‘index’라는 새로운 컬럼 명으로 추가합니다

### 데이터 Selection 및 Filtering

**DataFrame의 [ ] 연산자**  


In [45]:
titanic_df = pd.read_csv('train.csv')
print('단일 컬럼 데이터 추출:\n', titanic_df[ 'Pclass' ].head(3))
print('\n여러 컬럼들의 데이터 추출:\n', titanic_df[ ['Survived', 'Pclass'] ].head(3))
print('[ ] 안에 숫자 index는 KeyError 오류 발생:\n', titanic_df[0])

단일 컬럼 데이터 추출:
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 컬럼들의 데이터 추출:
    Survived  Pclass
0         0       3
1         1       1
2         1       3


KeyError: 0

넘파이에서 [ ] 연산자는 행의 위치, 열의 위치, 슬라이싱 범위 등을 지정해 데이터를 가져올 수 있었습니다.
하지만 DataFrame 바로 뒤에 있는 ‘[ ]’ 안에 들어갈 수 있는 것은 컬럼 명 문자(또는 컬럼 명의 리스트
객체), 또는 인덱스로 변환 가능한 표현식입니다.

In [47]:
titanic_df[0:2]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

앞에서 DataFrame의 [ ] 내에 숫자 값을 입력할 경우 오류가 발생한다고 했는데, Pandas의 Index 형태로 변환가능한 표현식은 [ ] 내에 입력할 수 있습니다.  
가령 titanic_df의 처음 2개 데이터를 추출하고자  titanic_df [ 0:2 ] 와 같은 슬라이싱을 이용하였다면 정확히 원하는 결과를 반환해 줍니다. 

In [48]:
titanic_df[ titanic_df['Pclass'] == 3].head(3)

PassengerId  Survived  Pclass                      Name     Sex   Age  \
0            1         0       3   Braund, Mr. Owen Harris    male  22.0   
2            3         1       3    Heikkinen, Miss. Laina  female  26.0   
4            5         0       3  Allen, Mr. William Henry    male  35.0   

   SibSp  Parch            Ticket   Fare Cabin Embarked  
0      1      0         A/5 21171  7.250   NaN        S  
2      0      0  STON/O2. 3101282  7.925   NaN        S  
4      0      0            373450  8.050   NaN        S

[ ] 내에 조건식을 입력하여 불린 인덱싱을 수행할 수 있습니다.  
[ ] 내의 불린 인덱싱 기능은 원하는 데이터를 편리하게 추출해주므로 자주 사용되며 DataFrame[ 0:2 ]과 같은 슬라이싱 연산으로 데이터를 추출하는 방법은 사용하지 않는게 좋습니다.

**DataFrame iloc[ ] 연산자**  

In [49]:
data = {'Name': ['Chulmin', 'Eunkyung','Jinwoong','Soobeom'],
        'Year': [2011, 2016, 2015, 2015],
        'Gender': ['Male', 'Female', 'Male', 'Male']
       }
data_df = pd.DataFrame(data, index=['one','two','three','four'])
data_df


Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male

In [51]:
data_df.iloc[1,0]

'Eunkyung'

In [52]:
data_df.iloc[0,"Name"]

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

위치기반 인덱싱을 제공합니다.

**DataFrame loc[ ] 연산자**  

In [55]:
data_df

Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male

In [54]:
data_df.loc['one', 'Name']

'Chulmin'

In [56]:
data_df_reset.loc[0, 'Name']

NameError: name 'data_df_reset' is not defined

loc[ ]는 명칭 기반으로 데이터를 추출합니다. 따라서 행 위치에는 데이터프레임 인덱스값을, 열 위치에는 칼럼명을 입력해줍니다. index가 숫자형일 수 있기 때문에 명칭 기반이라고 무조건 문자열으 입력한다는 선입견을 가져서는 안됩니다. 예를 들어 reset_index( )로 1부터 시작하는 정수형 인덱스값으로 변경한다면 loc[1,'Name']이 가능합니다.

In [59]:
print(data_df,'\n')
print('위치기반 iloc slicing\n', data_df.iloc[0:1, 0],'\n')
print('명칭기반 loc slicing\n', data_df.loc['one':'two', 'Name'])

           Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male 

위치기반 iloc slicing
 one    Chulmin
Name: Name, dtype: object 

명칭기반 loc slicing
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object


보통 슬라이싱 기호는 시작 값 ~ 종료 값 -1까지의 범위를 의미하지만 loc[ ]에서 사용되는 슬라이싱 기호 ':'는 시작 값 ~ 종료 값까지의 값의 범위를 의미합니다. 이유는 명칭 기반 인덱싱의 특성 때문으로 명칭은 숫자 형이 아닐 수도 있기 때문에 -1를 할 수가 없습니다.

**불린 인덱싱(Boolean indexing)**